src/runtime/ptx_dev.ll

```diff
@@ -346,44 +346,68 @@ define weak_odr i32 @halide_ptx_trap() nounwind uwtable alwaysinline {
 }
 
 ; llvm doesn't expose dot product instructions as intrinsics
-define weak_odr i32 @dp4a_s32_s32(i32 %a, i32 %b, i32 %i) nounwind readnone alwaysinline {
-       %d = tail call i32 asm "dp4a.s32.s32    $0, $1, $2, $3;", "=r,r,r,r"(i32 %a, i32 %b, i32 %i) nounwind readnone
+define weak_odr i32 @dp4a_s32_s32(<4 x i8> %a, <4 x i8> %b, i32 %i) nounwind readnone alwaysinline {
+       %a_32 = bitcast <4 x i8> %a to i32
```

Comment:
Does this result in as-good codegen as before? Now we load something as an i32, bit-cast it to 4xi8, then bitcast it back to i32? Seems risky.

Reply:
I think it's pretty safe. The bitcasts will certainly simplify away, and I'd expect even quite a degree of shuffles to simplify away.

The generated PTX code in at least a few of the test cases in cuda_8_bit_gemm is *identical* to master, including the register names, which suggests to me that the optimized LLVM IR is identical (not just similar).

Reply:
I've now verified that the codegen is identical for all of these tests.